src/Core/Data/AbstractTypedCollection.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * 2007-2019 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Data;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use PrestaShop\PrestaShop\Core\Exception\InvalidArgumentException;
+
+/**
+ * Class AbstractTypedCollection is an abstract collection class which checks
+ * that the inserted elements match the requested type.
+ */
+abstract class AbstractTypedCollection extends ArrayCollection
```

Comment:
wouldnt it be better to have `TypedCollection::__construct(string $type, array $elements = [])`? so we wouldnt need `abstract` at all?

Reply:
you're right it would work, but it changes the constructor prototype
even though you could override it again in the child class and remove this argument, the advantage I see with the abstract is that it forces you to implement it, you don't have a choice
whereas you could "forget" to override the constructor

Reply:
actually what i mean is that override would become optional. you could just do something like:

```php
$objects = new TypeCollection(MyObject::class);

$objects->add(new MyClass()); // all good
$objects->add(new SomeOtherClass()); // woops, type is not supported
```

Reply:
ok, got it. :)